perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+
+def MultiStreamWrites(storage_schema, host_to_connect=None):
+  """Run multi-stream write threads."""
+  RunMultiThreadedWorkersWriteResults(
+      WriteWorker,
+      (storage_schema,
+       host_to_connect,
+       FLAGS.object_size_kb,
+       FLAGS.objects_per_stream,
+       FLAGS.start_time),
+      storage_schema,
+      host_to_connect=host_to_connect)
+
+
+def MultiStreamReads(storage_schema, host_to_connect=None):
+  """Run multi-stream read threads."""
+  RunMultiThreadedWorkersWriteResults(
+      ReadWorker,
+      (storage_schema,
+       host_to_connect,
+       FLAGS.object_size_kb,
+       FLAGS.start_time),
+      storage_schema,
+      host_to_connect=host_to_connect)
+
+
+def SleepUntilTime(when, too_late=None):
+  """Sleep until a given time.
+
+  Args:
+    when: float. The time to sleep to, as a POSIX timestamp.
+
+    too_late: None or Exception. If when is before now + MIN_SLEEP,
+      raise the given exception. If too_late is None, log a message
+      but do not raise an exception.
+  """
+
+  now = time.time()
+  sleep_time = when - now
+  if sleep_time > MIN_SLEEP:
+    time.sleep(sleep_time)
+  elif too_late is not None:
+    raise too_late('Sleep time %s was too small' % sleep_time)
+  else:
+    logging.info('Sleep time %s was too small', sleep_time)
+
+
+OBJECTS_WRITTEN_FILE_PREFIX = 'pkb-objects-written-'
+
+
+def WriteWorker(storage_schema, host_to_connect,
+                object_size_kb, num_objects,
+                start_time, result_queue, worker_num):
+  payload_bytes, payload_string = GenerateWritePayload(
+      object_size_kb * BYTES_PER_KILOBYTE)
+
+  objects_written = []
+  start_times = []
+  latencies = []
+
+  if start_time is not None:
+    SleepUntilTime(start_time, too_late=Exception)
+
+  WriteObjectsFromBuffer(storage_schema, FLAGS.bucket,
+                         ('pkb_write_worker_%s' % worker_num),
+                         num_objects, payload_bytes, payload_string,
+                         object_size_kb * BYTES_PER_KILOBYTE,
+                         objects_written,
+                         object_start_times=start_times,
+                         latency_results=latencies,
+                         host_to_connect=host_to_connect)
+
+  objects_filename = os.path.join(TMP_DIR,
+                                  OBJECTS_WRITTEN_FILE_PREFIX + str(worker_num))
+
+  # Error handling deliberately omitted - abort on any exception.
+  if os.path.exists(objects_filename):
+    os.remove(objects_filename)
+  with open(objects_filename, 'w') as out:
+    json.dump(objects_written, out)
+
+  result_queue.put({'kind': 'upload',
+                    'start_times': start_times,
+                    'latencies': latencies,
+                    'stream_num': worker_num})
+
+
+def ReadWorker(storage_schema, host_to_connect,
+               object_size_kb, start_time,
+               result_queue, worker_num):
+
+  objects_filename = os.path.join(TMP_DIR,
```

Comment:
I ended up changing the logic so that the controlling thread writes one log file for all workers, rather than each individual worker saving its own file. Since the controlling process gets all the result data anyway to pass back to PKB, it's not really any more information flow, and it makes the log file logic much simpler. It also makes the worker thread functions small, which I like.